File: src/pusher.rs
/****************************************************************************
 * Copyright (C) 2020 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 ***************************************************************************/

use crate::api;
use crate::errors::GenericError;
use gio;
use glib;
use log::info;
use ostree;
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::fs::File;
use std::io;
use std::path::Path;
use std::path::PathBuf;
use std::result::Result;

pub struct Pusher {
    repo: ostree::Repo,
    repo_path: PathBuf,
    branches: HashMap<String, String>,
}

impl Pusher {
    pub fn new(repo_path: &Path, refs: Option<Vec<&str>>) -> Result<Pusher, glib::Error> {
        // Open repository
        let cancellable = gio::Cancellable::new();
        let repo = ostree::Repo::new_for_path(&repo_path);
        repo.open(Some(&cancellable))?;

        // Enumerate branches to push
        let mut branches = HashMap::new();
        if refs.is_some() {
            for refspec in refs.unwrap() {
                let rev = repo.resolve_rev(&refspec, false)?;
                branches.insert(refspec.to_string(), rev.to_string());
            }
        } else {
            for (branch, rev) in repo.list_refs(None, Some(&cancellable))? {
                branches.insert(branch, rev);
            }
        }

        Ok(Pusher {
            repo: repo,
            repo_path: repo_path.to_owned(),
            branches: branches,
        })
    }

    fn ostree_object_path(&self, object_name: &str) -> PathBuf {
        self.repo_path
            .join("objects")
            .join(&object_name[..2])
            .join(&object_name[2..])
    }

    fn needed_commits(
        &self,
        remote_rev: &str,
        local_rev: &str,
        commits: &mut Vec<String>,
    ) -> Result<(), GenericError> {
        let mut parent = Some(local_rev.to_string());
        let remote = if remote_rev == api::REV_NULL {
            None
        } else {
            Some(remote_rev.to_string())
        };

        while parent != remote {
            let parent_str = parent.unwrap().to_string();

            commits.push(parent_str.to_owned());

            match self
                .repo
                .load_variant_if_exists(ostree::ObjectType::Commit, &parent_str)
            {
                Err(_) => {
                    return Err(GenericError::new(&format!(
                        "Shallow history from commit {} doesn't contain remote commit {}",
                        &local_rev, &remote_rev
                    )));
                }
                Ok(commit) => {
                    parent = ostree::commit_get_parent(&commit).map(|s| s.as_str().to_string());
                    if parent.is_none() {
                        break;
                    }
                }
            }
        }

        if remote.is_some() && parent != remote {
            return Err(GenericError::new(&format!(
                "Remote commit {} not descendent of commit {}",
                &remote_rev, &local_rev
            )));
        }

        Ok(())
    }

    fn calculate_checksum(&self, path: &Path) -> io::Result<String> {
        let mut file = File::open(&path)?;
        let mut hasher = Sha256::new();
        io::copy(&mut file, &mut hasher)?;
        let hash = hasher.result();
        let hex = hash.as_ref().iter().map(|b| format!("{:x}", b)).collect();
        Ok(hex)
    }

    fn needed_objects(&self, revs: &Vec<String>) -> Result<Vec<api::NeededObject>, GenericError> {
        let mut objects: Vec<api::NeededObject> = Vec::new();
        let cancellable = gio::Cancellable::new();

        for rev in revs {
            match self.repo.traverse_commit(&rev, 0, Some(&cancellable)) {
                Err(_) => {
                    break;
                }
                Ok(reachable) => {
                    for object in reachable {
                        if let Some(object_name) =
                            ostree::object_to_string(object.checksum(), object.object_type())
                        {
                            match object.object_type() {
                                ostree::ObjectType::File => {
                                    // Make this a filez since we're archive-z2
                                    let file_object_name = format!("{}z", &object_name).to_string();
                                    let file_object_path =
                                        self.ostree_object_path(&file_object_name);
                                    let checksum = self
                                        .calculate_checksum(&file_object_path)
                                        .map_err(|e| GenericError::new(&format!("{}", e)))?;
                                    objects.push(api::NeededObject {
                                        rev: object.checksum().to_string(),
                                        object_name: file_object_name.to_owned(),
                                        object_path: file_object_path,
                                        checksum: checksum,
                                    });
                                }
                                ostree::ObjectType::Commit => {
                                    // Add in detached metadata
                                    let meta_object_name = &format!("{}meta", &object_name);
                                    let meta_object_path =
                                        self.ostree_object_path(&meta_object_name);
                                    let checksum = self
                                        .calculate_checksum(&meta_object_path)
                                        .map_err(|e| GenericError::new(&format!("{}", e)))?;
                                    if meta_object_path.exists() {
                                        objects.push(api::NeededObject {
                                            rev: object.checksum().to_string(),
                                            object_name: meta_object_name.to_owned(),
                                            object_path: meta_object_path,
                                            checksum: checksum,
                                        });
                                    }
                                }
                                _ => {
                                    let object_path = self.ostree_object_path(&object_name);
                                    let checksum = self
                                        .calculate_checksum(&object_path)
                                        .map_err(|e| GenericError::new(&format!("{}", e)))?;
                                    objects.push(api::NeededObject {
                                        rev: object.checksum().to_string(),
                                        object_name: object_name.to_owned(),
                                        object_path: object_path,
                                        checksum: checksum,
                                    });
                                }
                            }
                        }
                    }
                }
            }
        }

        Ok(objects)
    }

    pub fn check_update(
        &self,
        remote_refs: HashMap<String, String>,
    ) -> HashMap<String, (String, String)> {
        let mut update_refs: HashMap<String, (String, String)> = HashMap::new();
        let null_rev = api::REV_NULL.to_string();
        for (branch, rev) in &self.branches {
            let remote_rev = remote_refs.get(branch).unwrap_or(&null_rev);
            if rev != remote_rev {
                update_refs.insert(
                    branch.to_string(),
                    (remote_rev.to_string(), rev.to_string()),
                );
            }
        }

        update_refs
    }

    pub fn retrieve(
        &self,
        update_refs: &HashMap<String, (String, String)>,
    ) -> Result<Vec<api::NeededObject>, GenericError> {
        let mut commits: Vec<String> = Vec::new();
        for (branch, revs) in update_refs {
            info!(
                "Updating branch {} from {} to {}",
                &branch, &revs.0, &revs.1
            );
            self.needed_commits(&revs.0, &revs.1, &mut commits)?;
        }

        info!("Enumerating objects to send...");
        let objects = self.needed_objects(&commits)?;

        Ok(objects)
    }

    pub fn prune(&self) -> Result<(), GenericError> {
        let cancellable = gio::Cancellable::new();
        let (found, pruned, size) = self
            .repo
            .prune(ostree::RepoPruneFlags::None, -1, Some(&cancellable))
            .map_err(|e| GenericError::new(&format!("{}", e)))?;
        info!(
            "Pruned {}/{} objects, {} bytes deleted",
            pruned, found, size
        );

        Ok(())
    }
}

File: internal/push/pusher.go
// SPDX-FileCopyrightText: 2020 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

package push

import (
	"fmt"
	"os"

	"github.com/lirios/ostree-upload/internal/common"
	"github.com/lirios/ostree-upload/internal/logger"
	"github.com/lirios/ostree-upload/internal/ostree"
)

// Pusher allows you to push missing objects to an OSTree repository
type Pusher struct {
	repo     *ostree.Repo
	branches map[string]string
}

// NewPusher creates a new Pusher object
func NewPusher(repoPath string, refs []string) (*Pusher, error) {
	// Check if the repository path exist
	repo, err := ostree.OpenRepo(repoPath)
	if err != nil {
		return nil, err
	}

	// Enumerate branches to push
	branches := map[string]string{}
	if len(refs) == 0 {
		revisions, err := repo.ListRevisions()
		if err != nil {
			return nil, err
		}

		for branch, rev := range revisions {
			branches[branch] = rev
		}
	} else {
		for _, ref := range refs {
			rev, err := repo.ResolveRev(ref)
			if err != nil {
				return nil, err
			}

			branches[ref] = rev
		}
	}

	return &Pusher{repo, branches}, nil
}

// FindNeededCommits finds the commits of the local repository that the remove one doesn't have
func (p *Pusher) FindNeededCommits(remoteRev, localRev string) ([]string, error) {
	commits := []string{}

	parent := localRev

	for parent != remoteRev {
		logger.Debugf("Parent commit %s", parent)
		commits = append(commits, parent)

		newParent, err := p.repo.GetParentRev(parent)
		if err != nil {
			return nil, err
		}

		if newParent == "" {
			break
		} else {
			parent = newParent
		}
	}

	if remoteRev != "" && parent != remoteRev {
		return nil, fmt.Errorf("remote commit %v not descendent of commit %v", remoteRev, localRev)
	}

	return commits, nil
}

// FindObjectsForCommits finds the objects corresponding to the revisions that needs to be pushed to the receiver
func (p *Pusher) FindObjectsForCommits(revs []string) (common.Objects, error) {
	objects := make(common.Objects, 1024)

	for _, rev := range revs {
		revObjects, err := p.repo.TraverseCommit(rev, 0)
		if err != nil {
			return nil, err
		}

		for _, objectName := range revObjects {
			path := p.repo.GetObjectPath(objectName)
			if _, err := os.Stat(path); err != nil {
				return nil, err
			}

			checksum, err := common.CalculateChecksum(path)
			if err != nil {
				return nil, err
			}

			object := common.Object{Rev: rev, ObjectName: objectName, ObjectPath: path, Checksum: checksum}
			objects[objectName] = object
		}

	}

	return objects, nil
}

// CheckUpdate returns a map whose key is a branch and the value contains the corresponding
// revision in the remote and local repositories
func (p *Pusher) CheckUpdate(remoteRefs map[string]string) (map[string]common.RevisionPair, error) {
	updateRefs := make(map[string]common.RevisionPair)

	for branch, rev := range p.branches {
		remoteRev := remoteRefs[branch]
		if rev != remoteRev {
			updateRefs[branch] = common.RevisionPair{Server: remoteRev, Client: rev}
		}
	}

	return updateRefs, nil
}

// Prune prunes the repository
func (p *Pusher) Prune() error {
	total, pruned, size, err := p.repo.Prune(false, false)
	if err != nil {
		return err
	}

	logger.Infof("Pruned %d/%d objects, %d bytes deleted", pruned, total, size)

	return nil
}

// FindObjectsToPush finds which objects need to be pushed
func (p *Pusher) FindObjectsToPush(updateRefs map[string]common.RevisionPair) (common.Objects, error) {
	var commits []string

	for branch, revs := range updateRefs {
		logger.Actionf("Finding commits on branch \"%s\"...", branch)
		neededCommits, err := p.FindNeededCommits(revs.Server, revs.Client)
		if err != nil {
			return nil, err
		}
		commits = append(commits, neededCommits...)
	}

	logger.Action("Enumerating objects to send (this might take a while)...")
	neededObjects, err := p.FindObjectsForCommits(commits)
	if err != nil {
		return nil, err
	}

	return neededObjects, nil
}
